pkg/crd/crds_test.go

```diff
@@ -0,0 +1,238 @@
+			wantErr:  false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := validateCustomResourceYaml(tc.manifest, schemaValidator)
+			if tc.wantErr != (err != nil) {
+				t.Errorf("wantErr: %v, error, %s", tc.wantErr, err)
+			} else if err != nil {
+				t.Logf("Got expected validation error, %s", err)
+			}
+		})
+	}
+}
+
+func validateCustomResourceYaml(data string, vd validation.SchemaValidator) error {
+	cr := &unstructured.Unstructured{}
+	if err := yaml.Unmarshal([]byte(data), cr); err != nil {
+		return err
+	}
+	if err := validation.ValidateCustomResource(nil, cr.Object, vd).ToAggregate(); err != nil {
+		return fmt.Errorf("%v %v is invalid, err: %w", cr.GetKind(), cr.GetName(), err)
+	}
+	return nil
+}
+
+var bp_bad_matchLabels_value = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchLabels:
+      name: edgeplatform-test-wec1, edgeplatform-test-wec2
+  downsync:
+  - objectSelectors:
+    - matchLabels:
+        app.kubernetes.io/part-of: wec-gpu
+  wantSingletonReportedState: true
+`
+
+var bp_good_matchLabels = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchLabels: {"sub.do-main.test/0.9-5_Xy": "This-is_1.test"}
+  downsync:
+  - objectSelectors:
+    - matchLabels:
+        app.kubernetes.io/part-of: wec-gpu
+  wantSingletonReportedState: true
+`
+
+var bp_bad_matchExpressions_key = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchExpressions:
+    - key: -model
+      operator: In
+      values:
+      - H200
+      - H100
+  downsync:
+  - objectSelectors:
+    - matchLabels:
+        app.kubernetes.io/part-of: wec-gpu
+  wantSingletonReportedState: true
+`
+
+var bp_bad_matchExpressions_op = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchExpressions:
+    - key: model
+      operator: OutOf
+      values:
+      - H200
+      - H100
+  downsync:
+  - objectSelectors:
+    - matchLabels:
+        app.kubernetes.io/part-of: wec-gpu
+  wantSingletonReportedState: true
+`
+var bp_bad_matchExpressions_value = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchExpressions:
+    - key: model
+      operator: In
+      values:
+      - H200
+      - H100,A100
+  downsync:
+  - objectSelectors:
+    - matchLabels:
+        app.kubernetes.io/part-of: wec-gpu
+  wantSingletonReportedState: true
+`
+
+var bp_good_matchExpressions = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchExpressions:
+    - key: model
```

Comment:
I recommend a more challenging key. See what https://v1-28.docs.kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set says about allowable label names.

Reply:
Done.